internal/runtime/client/client.go

```diff
@@ -74,180 +92,247 @@ func (c *client) Extension(ext *runtimev1.Extension) ExtensionClient {
+			return err
+		}
+		// Follow-up: Call all extensions irrespective of error. Don't short-circuit on an error.
+		if err := h.CallExtension(ctx, registration.Name, request, response); err != nil {
 			return err
 		}
+		responses = append(responses, response)
 	}
-
+	h.aggregateResponses(responses, response)
 	return nil
 }
 
-func createHttpClient(registration *registry.RuntimeExtensionRegistration) httpClient {
-	return httpClient{}
-}
-
-type httpClient struct {
+func (h *hookClient) aggregateResponses(list []runtime.Object, into runtime.Object) {
+	panic("implement a response aggregation mechanism")
 }
 
-func (h *httpClient) Call(ctx context.Context, in, out runtime.Object) error {
+func (h *hookClient) CallExtension(ctx context.Context, name string, request, response runtime.Object) error {
+	registration, err := h.client.registry.Get(name)
+	if err != nil {
+		return err
+	}
+	opts := &httpCallOptions{
+		catalog:       *h.client.catalog,
+		config:        registration.ClientConfig,
+		gvh:           registration.GroupVersionHook,
+		name:          registration.Name,
+		timeout:       registration.TimeoutSeconds,
+		failurePolicy: registration.FailurePolicy,
+	}
+	if err := httpCall(ctx, request, response, opts); err != nil {
+		return errors.Wrap(err, "failed to call extension")
+	}
 	return nil
 }
 
-type ServiceOption interface {
-	ApplyToServiceOptions(*ServiceOptions)
-}
+type ExtensionClient interface {
+	// Discover makes the discovery call on the extension and updates the runtime extensions
+	// information in the extension status.
+	// TODO: Need a final decision on if we also want to run register inside discover.
+	Discover() (*runtimev1.Extension, error)
 
-type SpecVersion string
+	// Register registers the extension with the manager.
+	Register() error
 
-func (o SpecVersion) ApplyToServiceOptions(opts *ServiceOptions) {
-	opts.Version = string(o)
+	//Unregister unregisters the extension with the manager.
+	Unregister() error
 }
 
-type ServiceOptions struct {
-	Version string // TODO: pointer?
+var _ ExtensionClient = &extensionClient{}
+
+type extensionClient struct {
+	client *client
+	ext    *runtimev1.Extension
 }
 
-type ServiceClient interface {
-	Invoke(ctx context.Context, in, out runtime.Object) error
+func (e *extensionClient) Discover() (*runtimev1.Extension, error) {
+	gvh, err := e.client.catalog.GroupVersionHook(hooksv1alpha1.Discovery)
+	if err != nil {
+		return nil, err
+	}
+
+	request := &hooksv1alpha1.DiscoveryHookRequest{}
+	response := &hooksv1alpha1.DiscoveryHookResponse{}
+
+	// Future work: The discovery runtime extension could be operating on a different hook version than
+	// the latest. We will have to loop through different versions of the discover hook here to actually
+	// finish discovery.
+	opts := &httpCallOptions{
+		catalog: *e.client.catalog,
+		config:  e.ext.Spec.ClientConfig,
+		gvh:     gvh,
+	}
+	if err := httpCall(context.Background(), request, response, opts); err != nil {
```

Comment:
Let's please already pass through the context (i.e. add a ctx parameter to this func).